File: internal/config/config_reality_test.go
package config

import (
	"os"
	"path/filepath"
	"testing"
)

// TestLoad_RealConfig verifies that the actual .nightgauge/config.yaml
// in this repository parses correctly. This test fails if the config format
// changes break the parser — providing an early warning before CI or users
// are affected.
//
// Skip conditions (graceful degradation):
//   - Config file not found (shallow CI clone, running from a fork)
//   - File unreadable (permissions issue)
func TestLoad_RealConfig(t *testing.T) {
	// Navigate from internal/config/ to repo root (two levels up).
	// This uses a relative path so it works regardless of GOPATH or checkout
	// location, matching the pattern established by fixtureYAML().
	repoRoot, err := filepath.Abs(filepath.Join("..", ".."))
	if err != nil {
		t.Skipf("could not resolve repo root: %v", err)
	}

	configPath := filepath.Join(repoRoot, ".nightgauge", "config.yaml")
	if _, err := os.Stat(configPath); os.IsNotExist(err) {
		t.Skipf("real config.yaml not found at %s — skipping reality test", configPath)
	}

	cfg, err := LoadYAML(configPath)
	if err != nil {
		t.Fatalf("LoadYAML real config: %v", err)
	}

	// Verify owner — this is the most critical field; an empty owner means
	// GitHub API calls will fail silently.
	if cfg.Owner == "" {
		t.Error("Owner is empty — config format may have changed")
	}
	if cfg.Owner != "nightgauge" {
		t.Errorf("Owner = %q, want %q", cfg.Owner, "nightgauge")
	}

	// The public-core dogfood config must not route a clone into an organization
	// Project. Operators configure a board locally when they opt into board-backed
	// commands.
	if cfg.ProjectNumber != 0 {
		t.Errorf("ProjectNumber = %d, want 0 for safe public defaults", cfg.ProjectNumber)
	}

	// Verify repo name is populated (hybrid format top-level field).
	if cfg.DefaultRepo == "" {
		t.Error("DefaultRepo (repo) is empty — hybrid format top-level repo field not parsed")
	}
	if cfg.DefaultRepo != "nightgauge" {
		t.Errorf("DefaultRepo = %q, want %q", cfg.DefaultRepo, "nightgauge")
	}
}

File: packages/nightgauge-vscode/tests/config/configReality.test.ts
/**
 * Reality tests for .nightgauge/config.yaml
 *
 * These tests load the actual config file from this repository and verify that
 * the parser and path resolver handle its hybrid format correctly. They fail
 * when format changes break the parser — providing an early warning before
 * CI or users are affected.
 *
 * Skip condition: config file absent (shallow CI clone, fork without config).
 *
 * @see Issue #1937 - Config reality tests: verify actual config.yaml parses correctly
 */

import { describe, it, expect } from "vitest";
import * as path from "path";
import { existsSync } from "fs";
import {
  resolveConfigPath,
  getRepoIdentity,
  CONFIG_DIR,
  CONFIG_FILE_NAME,
} from "../../src/utils/configPathResolver";

// Resolve repo root from vitest CWD (packages/nightgauge-vscode/)
const repoRoot = path.resolve(process.cwd(), "..", "..");
const configPath = path.join(repoRoot, CONFIG_DIR, CONFIG_FILE_NAME);
const configExists = existsSync(configPath);

describe.skipIf(!configExists)("Config reality tests — .nightgauge/config.yaml", () => {
  it("resolveConfigPath finds the real config file as non-legacy", async () => {
    const result = await resolveConfigPath(repoRoot);

    expect(result.exists).toBe(true);
    expect(result.isLegacy).toBe(false);
    expect(result.path).toBe(configPath);
  });

  it("getRepoIdentity reads owner and repo from real config", async () => {
    const identity = await getRepoIdentity(repoRoot);

    expect(identity).not.toBeNull();
    expect(identity!.owner).toBe("nightgauge");
    expect(identity!.repo).toBe("nightgauge");
  });
});
